src/lib.rs

```diff
@@ -336,6 +351,7 @@ impl LockFile {
             panic!("Cannot lock if already owning a lock");
         }
         sys::lock(self.desc)?;
+        fileid::take_lock(self.id);
```

Comment:
I believe there is a potential race condition here.

I'd rather have the system lock being more "internal" than the `fildeid` lock. What I mean is, let's lock `fileid` before locking via `sys`, and let's `sys` before releasing `fileid`. Consider the following situation:

1. Thread A acquires both locks, and now runs with their acquired lock.
2. Thread A decides to release the lock, and first unlocks `fileid`.
3. Thread B executes the system lock with sucess, because we are in the same process.
4. Thread A unlocks the system level lock.
5. Thread B acquires the `fileid` lock.

Now, thread B believes the file is locked, but it isn't. How to change this? Invert the order. If we invert the order of the locks, the situation would happen as follows:

1. Thread A acquires both locks, and now runs with their acquired lock.
2. Thread A decides to release the lock, and first unlocks OS-level lock.
3. Thread B needs to lock `fileid`, but it will have to wait thread A unlock it.
4. Thread A unlocks the `fileid` level lock.
5. Thread B acquires the `fileid` lock.
6. Thread B acquires OS-level lock.

```suggestion
        fileid::take_lock(self.id);
        if let Err(error) = sys::lock(self.desc) {
            fileid::release_lock(self.id);
            Err(error)?;
        }
```

Reply:
This is a very good catch! I can't remember why I decided to do it in the order I did, but I think that your approach is correct!  The only thing I'd suggest in addition would be a comment explaining this in the file.